### PR TITLE
Adding the Nodekeeper doc

### PR DESCRIPTION
**Pull request by anjoshi24**
Signed-off-by: Aniket Joshi <anjoshi@redhat.com>

### What type of PR is this?
_documentation_

### What this PR does / why we need it?
PR adds a detailed documentation for the Nodekeeper Controller as need according to [OSD-3472](https://issues.redhat.com/browse/OSD-3472)

### Which Jira/Github issue(s) this PR fixes?
[OSD-3472](https://issues.redhat.com/browse/OSD-3472)

_Fixes #_

### Pre-checks (if applicable):
- [ ] Tested latest changes against a cluster
- [ ] Ran `make generate` command locally to validate code changes
- [ ] Included documentation changes with PR

### CODE REVIEW

File: docs/nodekeeper.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+The `Nodekeeper` controller will flag through the `metricNodeDrainFailed` metric generating alerts to SRE-P team on-call that any worker node continue to unsuccessfully drain in spite of `NodeDrain` strategies.
+
+## Inside the Nodekeeper Controller
+- The `Nodekeeper` controller mechanism starts with setting the controller with [Add](https://github.com/openshift/managed-upgrade-operator/blob/master/pkg/controller/nodekeeper/nodekeeper_controller.go#L33) function which add sets up the controller for the operator manager - setting up a watch on the resource. For the `Nodekeeper` controller the mechanism is setting up a watch on the resource of Node.
+
+- Now as already specified, the `Nodekeeper` controller works only towards the worker nodes so for keeping the master nodes aside, an [IgnoreMasterPredicate](https://github.com/openshift/managed-upgrade-operator/blob/master/pkg/controller/nodekeeper/ignoremaster_predicate.go) is used which makes sure that the controller only targets worker nodes in it's mechanism.
```
Comment (dofinn) [inline docs/nodekeeper.md:12]: Should we mention how Infra nodes are handled?

File: docs/nodekeeper.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+- The controller mechanism checks for if a node is cordoned inside the `Reconcile` function at [IsNodeCordoned](https://github.com/openshift/managed-upgrade-operator/blob/master/pkg/controller/nodekeeper/nodekeeper_controller.go#L130) based on [node.go/IsNodeCordoned](https://github.com/openshift/managed-upgrade-operator/blob/367c5754bd0267f3b6dfd782f30f52a3dc94a3c8/pkg/machinery/node.go#L13). The [node.go/IsNodeCordoned](https://github.com/openshift/managed-upgrade-operator/blob/367c5754bd0267f3b6dfd782f30f52a3dc94a3c8/pkg/machinery/node.go#L13) function checks for an `Unschedulable` and `Tainted` node especially for `TaintEffectNoSchedule` node and if it's matched then sets the node as cordoned and also saves the timestamp at which the node was found cordoned.
```
Comment (dofinn) [inline docs/nodekeeper.md:17]: "node especially" = "node specifically" ?

File: docs/nodekeeper.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+If the node is found to be cordoned, the controller acts on the `drainStrategy` for the node and if it fails to fix the node then sets `metricNodeDrainFailed` metric and alerts SRE-P and then if the node is no longer cordoned then resets the metric using `ResetMetricNodeDrainFailed`.
+
+- Next the controller is looking at the nodes in the state of upgrading and running workloads disrupting the upgrade such as [Pod Disruption Budgets](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets) or `finalizers` as the nodes should drain the workloads to continue the upgrade. In the mechanism the `NodeDrain.Timeout` is looked at around a node if it does not drain during specified period, we need to act upon the `drainStrategy`.
```
Comment (dofinn) [inline docs/nodekeeper.md:20]: > In the mechanism the `NodeDrain.Timeout` is looked at around a node if it does not drain during specified period, we need to act upon the `drainStrategy`.
This sentence is not clear to me, could you please re-word?

**Review by dofinn — COMMENTED** [review]
Awesome write up @anjoshi24 . Few minor comments. Great job :+1

File: docs/nodekeeper.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Nodekeeper Controller
+
+## About
+The `managed-upgrade-operator` provides a mechanism for keeping track of upgrading worker nodes and seeks to ensure their timely and eventual upgrade.
+
+The `Nodekeeper` controller makes sure that if an upgrading worker node is facing difficulty draining due to conditions such as [Pod Disruption Budgets](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets) or stuck finalizers, then `Nodekeeper` controller will perform functions for driaining the nodes properly and subsequent upgrade continuation.
```
[the remark this answers is not in the document]
Comment (anjoshi24) [reply docs/nodekeeper.md]: Changes made according to the suggestion.

File: docs/nodekeeper.md (hunk 2 above)
Comment (anjoshi24) [reply docs/nodekeeper.md]: Changes made according to the suggestion.

**Pull request merged**